open-craft/opencraft
# Restrict access to /instance and /api endpoints

This change restricts access to the /instance and /api URLs to staff users, and users with `InstanceManagerPermission`.

**JIRA tickets**: Implements [OC-1550](https://tasks.opencraft.com/browse/OC-1550)

**Screenshots**: 

Non-privileged users see Permission Denied for /api URLs:

![screen shot 2016-06-08 at 4 30 54 pm](https://cloud.githubusercontent.com/assets/7556571/15885464/7feabef8-2d96-11e6-8aed-18221bb166d1.png)

**Testing instructions**:

1. Set up an opencraft devstack according to [instructions in README](https://github.com/open-craft/opencraft#install).
1. Be sure to run `make migrate`, as this change introduces a new permission model.
1. Run automatic tests: `make test`
1. Set up users for manual tests: `make shell`
```python
In [1]: from django.contrib.auth.models import User
In [2]: beta = User.objects.create_user('beta', 'beta@example.com', 'beta')
In [3]: staff = User.objects.create_user('staff', 'staff@example.com', 'edx')
In [4]: staff.is_staff = True
In [5]: staff.save()
In [6]: superuser = User.objects.create_user('superuser', 'superuser@example.com', 'super')
In [7]: superuser.is_superuser = True
In [8]: superuser.save()
```
1. With each user in [`beta`, `staff`, `superuser`]:
    * Visit [http://localhost:5000](http://localhost:5000), and click Login.
        * beta should redirect to `/registration/`
        * staff, superuser should redirect to `/instance/`
    * Visit [http://localhost:5000/instance/](http://localhost:5000/instance/):
        * beta should redirect to `/registration/`
        * staff, superuser should redirect to `/instance/`
    * Visit [http://localhost:5000/api/](http://localhost:5000/api/)
        * beta user should get 403
        * staff, superuser should get 200, and be able to navigate further
    * Click 'Logout' in between each user.
        * Should redirect to `/registration/`

**Author notes and concerns**:

1. `make test_prospector`'s `pylint` tests were returning a number of `no-member` errors for inherited members of the new `InstanceManagerPermission` and `InstanceManagerPermissionManager` classes.  Errors messages are visible in the [initial CircleCI run](https://circleci.com/gh/open-craft/opencraft/1083). 

   This appears to be a [pylint-django bug, addressed by PR #74](https://github.com/landscapeio/pylint-django/pull/74).  I've suppressed the errors with commit [efcbac2](https://github.com/open-craft/opencraft/pull/104/commits/efcbac2e30c249538e31e1e6ac87ac32f8964f04).

**Reviewers**
